packages/repository-tests/src/crud/transactions.suite.ts

```diff
@@ -0,0 +1,133 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/repository-tests
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  Entity,
+  IsolationLevel,
+  juggler,
+  model,
+  property,
+  Transaction,
+  TransactionalRepository,
+} from '@loopback/repository';
+import {expect, skipIf, toJSON} from '@loopback/testlab';
+import {Suite} from 'mocha';
+import {withCrudCtx} from '../helpers.repository-tests';
+import {
+  CrudFeatures,
+  CrudTestContext,
+  DataSourceOptions,
+  TransactionalRepositoryCtor,
+} from '../types.repository-tests';
+
+// Core scenarios for testing CRUD functionalities of Transactional connectors
+// Please keep this file short, put any advanced scenarios to other files
+export function transactionSuite(
+  dataSourceOptions: DataSourceOptions,
+  repositoryClass: TransactionalRepositoryCtor,
+  connectorFeatures: CrudFeatures,
+) {
+  skipIf<[(this: Suite) => void], void>(
+    !connectorFeatures.supportsTransactions,
+    describe,
+    `transactions`,
+    () => {
+      @model()
+      class Product extends Entity {
+        @property({
+          type: connectorFeatures.idType,
+          id: true,
+          generated: true,
+          description: 'The unique identifier for a product',
+        })
+        id: number | string;
+
+        @property({type: 'string', required: true})
+        name: string;
+
+        constructor(data?: Partial<Product>) {
+          super(data);
+        }
+      }
+
+      describe('create-retrieve with transactions', () => {
+        let repo: TransactionalRepository<Product, typeof Product.prototype.id>;
+        before(
+          withCrudCtx(async function setupRepository(ctx: CrudTestContext) {
+            repo = new repositoryClass(Product, ctx.dataSource);
+            await ctx.dataSource.automigrate(Product.name);
+          }),
+        );
+
```

Comment:
What happens when a test opens a transaction and aborts on an assertion error before it has a chance to commit or reject the transaction?

Let's make the tests more robust to ensure they always close the transactions, even on test failure.

- Add a new `let tx: Transaction` variable.
- Modify tests to store the transaction to this variable, instead of creating a new test-local `const tx`.
- Add an `beforeEach` hook to set `tx` to `undefined`
- Add an `afterEach` hook to call `tx.rollback()` if `tx` is still set
- Figure out how to handle the case where the test has committed or rolled back a transaction. To begin with, we can change the tests to set `tx = undefined` after it has been committed or rolled back. A better solution would be to allow the `afterEach` hook to find out whether the transaction is still active, and do nothing if it is not. Since this would require changes in juggler and/or loopback-connector, I am fine to leave such improvement out of scope of this PR.

Reply:
Sounds great, thank you @bajtos!

Reply:
>Figure out how to handle the case where the test has committed or rolled back a transaction. To begin with, we can change the tests to set tx = undefined after it has been committed or rolled back. A better solution would be to allow the afterEach hook to find out whether the transaction is still active, and do nothing if it is not. Since this would require changes in juggler and/or loopback-connector, I am fine to leave such improvement out of scope of this PR.

I added two more properties to the `Transaction` interface, one being `id` and the other `connection`. I don't have any good typing info for `connection` so I've left it as `any`. That way, if there is no connection, we know it's not active. On the other hand, I can add `ensureActive` method from https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/transaction.js#L197 to the type-def for `Transaction` and use that instead. Thoughts?

Reply:
`ensureActive` does not seem very useful for our use case - we don't want to throw & catch an error in order to detect whether a transaction is active.

Personally, I'd prefer the transaction to provide `isActive` property that returns a boolean.

A mock-up:

```ts
Object.defineProperty(Transaction.prototype, 'isActive', {
  get: function() { return !!this.connection }
}
```

(The code needs to be added to loopback-connector, then we need to add `isActive` property to type defs both in loopback-connector and LB4.)

Reply:
Created #3471